metadata/src/retain.rs

```diff
@@ -211,6 +211,10 @@ where
     }
     update_extrinsic_types(&mut metadata.extrinsic, &map_ids);
     update_type(&mut metadata.ty, &map_ids);
+
+    // The retain pallets method strips the runtime APIs information,
+    // as it only takes into account working with pallets.
+    metadata.apis = Default::default();
```

Comment:
Are runtime APIs distinct from pallets then? I had assumed that each pallet would bring in some of it's own runtime APIs?

Reply:
I'm not sure if we can confidently identify which pallets are bringing their own APIs.

For example, the `Metadata` [trait](https://github.com/paritytech/substrate/blob/471b49b09fd0d2a3057090ab0720f4e39faed5c5/primitives/api/src/lib.rs#L721) lives as a separate entity and users define and implement these with `decl_runtime_apis` and `impl_runtime_apis` which I don't think will bind the API to any specific pallet.

Although it might be possible to "identify" the pallet by trait name, similar to the [system pallet](https://github.com/paritytech/substrate/blob/471b49b09fd0d2a3057090ab0720f4e39faed5c5/bin/node-template/runtime/src/lib.rs#L461-L465). However, that might not be deterministic, in the sense that users can name their APIs different from the pallet that defines them. 

For the retain metadata, I removed the runtime API for simplicity.

We could add a flag on the CLI to also retain specified runtime traits similar to
```bash
retain --pallets System --runtime-apis Metadata
```

That would imply we add two filters to those methods:
- one filter for identifying the pallets
- another for identifying the runtime APIs: identify just the `trait_name` (ie `Metadata` instead of `Metadata_metadata` and other fns)

I think that would give users the most control over what they want to retain, let me know what you think of it 🙏

Reply:
Ah thanks for the info!

Let's add another issue to allow trimming of the runtime APIs in the CLI tool (we don't need to do it here; it's not super important but would be nice for people to be able to keep only the runtime APIs they care about, if any)

And indeed, trimming pallets and trimming Runtime APIs should be different filters then; triming pallets shouldn't also remove the runtime APIs and vice versa :)